### DISCUSSION TITLE
JBrowse 2 linear genome view React component; bigwig tracks

### DISCUSSION THREAD

**Discussion by arpanda** [GMOD/jbrowse-components#2380] [Q&A]
I was trying to visualize some data for phythopthora genomes using a react application. For that, I used the following two links as guide.
1. https://www.npmjs.com/package/@jbrowse/react-linear-genome-view
2. https://github.com/GMOD/jbrowse-components/blob/HEAD/products/jbrowse-react-linear-genome-view/docs/example.md

 Codes are available at https://github.com/computational-genomics-lab/eumicrobe-db-client/blob/master/src/jbrowse/jbrowse.js

Somehow, here the bigwig track code is not running and giving error, I had to comment that section of the code to run the script. Here is that section of the code.
```
{
    trackID: "SSR_phy",
    name: "SSR",
    assemblyNames:["Phyag_NZFS3770"],
    type: "QuantitativeTrack",
    adapter: {
            type: "BigWig",
            bigWigLocation: { uri: ssrbw },
    }
  },
```

I also tried the add track option to show the [same bigwig file](https://github.com/computational-genomics-lab/eumicrobe-db-client/blob/master/src/genome_data/Phyag_NZFS3770.ssr.bw) but failed. 

The code are temporarily hosted at http://97.107.139.167:3000/jbrowse/jbrowse  

Any suggestion to fix the issue.

**Comment by cmdcolin** [comment]
>I also tried the add track option to show the same bigwig file but failed.


what is the error you see?

Also, it is somewhat unusual to be "importing" the fa, fai, and bigwig
 like this

```
import fasta from '../genome_data/Phyag_NZFS3770.fna'

```

that behavior would be up to your particular build system how that is handled e.g.  it would require an advanced "loader" system from webpack or similar, because by default you generally just import js files

it does seem to be working for you with the fa and fai file (maybe because they are plaintext ?) but then it is failing on the bigwig, possibly because it is a binary format (just a guess...)

in any case, it may be good to not rely on this ability to import random files: you may want to make these setup on a fileserver somewhere, and refer to them by URL (as it is done in the tutorial)

**Reply by arpanda** [reply]
Here is the screenshot of the error, while trying to add bigwig track from the browser. 
<img width="848" alt="image" src="https://user-images.githubusercontent.com/21196859/136427715-1ac048f9-9fb7-4f87-9ac2-59a6574f7879.png">

> that behavior would be up to your particular build system how that is handled e.g. it would require an advanced "loader" system from webpack or similar, because by default you generally just import js files
> 
> it does seem to be working for you with the fa and fai file (maybe because they are plaintext ?) but then it is failing on the bigwig, possibly because it is a binary format (just a guess...)
> 

I also think the same but not sure how to debug or solve it. Any suggestion to debug or solve?

**Reply by cmdcolin** [reply]
Currently we cannot add tracks fro the embedded LGV in the browser. See this issue https://github.com/GMOD/jbrowse-components/issues/1634

**Reply by arpanda** [reply]
Probably, JBrowse will have that feature in future. 

Also, tracklist setup for BigWig track also not responding in the embedded LGV.  

> ```
> {
>     trackID: "SSR_phy",
>     name: "SSR",
>     assemblyNames:["Phyag_NZFS3770"],
>     type: "QuantitativeTrack",
>     adapter: {
>             type: "BigWig",
>             bigWigLocation: { uri: ssrbw },
>     }
>   },
> ```

**Reply by cmdcolin** [reply]
I believe the idea of loading the BW using import as you have done above may not be correct....it might be loading a large data URL or something similar, I am not sure. That could break it. I posted this comment before https://github.com/GMOD/jbrowse-components/discussions/2380#discussioncomment-1427847 the alternative would be to put the bw file in your static resources directory or other file hosting, and just give an absolute URL to the file

**Reply by arpanda** [reply]
To check if the bigwig track properly or not, I have added the below bigwig track with the [example config](https://github.com/GMOD/jbrowse-components/blob/HEAD/products/jbrowse-react-linear-genome-view/docs/example.md). 
```
  {
    trackID: "SSR_phy",
    name: "SSR",
    assemblyNames:["GRCh38"],
    type: "QuantitativeTrack",
    adapter: {
            type: "BigWig",
            bigWigLocation: { uri: 'https://www.encodeproject.org/files/ENCFF318QRJ/@@download/ENCFF318QRJ.bigWig'},
    }
  },
```
I also created a public url for that file and got same response.
The problem still persists.

**Reply by cmdcolin** [reply]
What is the error message?  I think you posted the error from adding in the browser but not from the config

**Reply by arpanda** [reply]
This time, I added in the track config. The browser consol is not reporting any error but the page is loading phase only. Here is temporary link for setup . http://97.107.139.167:3000/jbrowse/jbrowse

**Reply by cmdcolin** [reply]
The config should be trackId not trackID

**Reply by cmdcolin** [reply]
adapter.type should also be BigWigAdapter not BigWig

**Reply by cmdcolin** [reply]
There is a certain danger to hand-editing configs, our config system is very strict and there can be trouble if the config fails to validate the schema. You can see at the link eventually the error prints out 
```
`Uncaught Error: [mobx-state-tree] Error while converting  {"config":{"assembly":{"name":"GRCh38","sequence":{"type":"ReferenceSequenceTrack","trackId":"GRCh38-ReferenceSequenceTrack","adapter":{"type":"BgzipFastaAdapter","fastaLocation":{"uri":"https://s3.amazonaws.com/jbrowse.org/genomes/GRCh38/fasta/GRCh38.fa.gz","locationTyp......}]}]}}}` to `ReactLinearGenomeView`
```
 which indicates that track does not have the right config schema. fixing can be tricky so it is best to try to get it right the first time if possible, by using the jbrowse CLI or jbrowse admin-server to edit configs, even though these are not geared towards the embedded (@jbrowse/react-linear-genome-view) component currently.

**Reply by arpanda** [reply]
Got it. It's working now. Thanks Colin.
